File: turtoise_future/strategies/pair_trading/exit.py
"""Exit logic for pair trading"""

import json
from typing import List, Dict
from ...config.settings import settings
from ...data.cache import get_candles_recent, get_contract_cn_name
from .cointegration import calculate_zscore
from ...execution.broker import place_market_order


def load_open_positions() -> List[Dict]:
    """Load open positions from JSON file"""
    try:
        with open("bot_agents.json", encoding="utf-8") as f:
            return json.load(f)
    except FileNotFoundError:
        return []


def save_open_positions(bot_agents: List[Dict]) -> None:
    """Save open positions to JSON file"""
    with open("bot_agents.json", "w", encoding="utf-8") as f:
        json.dump(bot_agents, f, ensure_ascii=False)


def manage_trade_exits() -> str:
    """
    Manage exiting open positions based on exit rules.

    Returns:
        Status string
    """
    open_positions = load_open_positions()

    if len(open_positions) < 1:
        return "complete"

    save_output = []

    for position in open_positions:
        is_close = False

        position_market_m1 = position["market_1"]
        position_size_m1 = position["order_m1_size"]
        position_side_m1 = position["order_m1_side"]

        position_market_m2 = position["market_2"]
        position_size_m2 = position["order_m2_size"]
        position_side_m2 = position["order_m2_side"]

        series_1 = get_candles_recent(position_market_m1)
        series_2 = get_candles_recent(position_market_m2)

        if settings.close_at_zscore_cross:
            hedge_ratio = position["hedge_ratio"]
            z_score_traded = position["z_score"]

            if len(series_1) > 0 and len(series_1) == len(series_2):
                spread = series_1 - series_2 * hedge_ratio
                z_score_current = calculate_zscore(spread).values.tolist()[-1]

                z_score_level_check = abs(z_score_current) > settings.zscore_threshold
                z_score_cross_check = (z_score_current < 0 and z_score_traded > 0) or (
                    z_score_current > 0 and z_score_traded < 0
                )

                if z_score_level_check and z_score_cross_check:
                    is_close = True

        if is_close:
            side_m1 = "SELL" if position_side_m1 == "BUY" else "BUY"
            side_m2 = "SELL" if position_side_m2 == "BUY" else "BUY"

            accept_price_1 = series_1[-1]
            accept_price_2 = series_2[-1]

            try:
                print(f"Closing market 1: {get_contract_cn_name(position_market_m1)}")
                place_market_order(
                    market=position_market_m1,
                    side=side_m1,
                    size=position_size_m1,
                    price=accept_price_1,
                    reduce_only=True,
                )

                print(f"Closing market 2: {get_contract_cn_name(position_market_m2)}")
                place_market_order(
                    market=position_market_m2,
                    side=side_m2,
                    size=position_size_m2,
                    price=accept_price_2,
                    reduce_only=True,
                )
            except Exception as e:
                print(f"Exit failed: {e}")
        else:
            save_output.append(position)

    print(f"{len(save_output)} Items remaining. Saving file ...")
    save_open_positions(save_output)

    return "complete"

File: rust/src/strategies/exit.rs
//! Exit logic for pair trading

use crate::config::Args;
use crate::data::MarketPrices;
use crate::strategies::cointegration::calculate_zscore;
use crate::strategies::entry::{load_open_positions, save_open_positions};
use std::path::Path;

/// Manage exiting open positions based on exit rules
pub fn manage_trade_exits(
    args: &Args,
    bot_agents_path: &Path,
    prices: &MarketPrices,
) -> Result<String, String> {
    let open_positions = load_open_positions(bot_agents_path);

    if open_positions.is_empty() {
        return Ok("complete".to_string());
    }

    let mut save_output = Vec::new();

    for position in &open_positions {
        let mut is_close = false;

        let series_1 = prices.get_series(&position.market_1);
        let series_2 = prices.get_series(&position.market_2);

        if args.close_at_zscore_cross {
            let hedge_ratio = position.hedge_ratio;
            let z_score_traded = position.z_score;

            if !series_1.is_empty() && series_1.len() == series_2.len() {
                let spread: Vec<f64> = series_1
                    .iter()
                    .zip(series_2.iter())
                    .map(|(s1, s2)| s1 - hedge_ratio * s2)
                    .collect();

                let z_scores = calculate_zscore(&spread, args.window);
                let z_score_current = z_scores.last().copied().unwrap_or(f64::NAN);

                if !z_score_current.is_nan() {
                    let z_score_level_check = z_score_current.abs() > args.zscore_threshold;
                    let z_score_cross_check = (z_score_current < 0.0 && z_score_traded > 0.0)
                        || (z_score_current > 0.0 && z_score_traded < 0.0);

                    if z_score_level_check && z_score_cross_check {
                        is_close = true;
                    }
                }
            }
        }

        if is_close {
            // In a real implementation, this would place closing orders
            // For now, we just log the exit
            println!(
                "Closing position: {}/{}",
                position.market_1, position.market_2
            );
        } else {
            save_output.push(position.clone());
        }
    }

    println!("{} Items remaining. Saving file ...", save_output.len());
    save_open_positions(bot_agents_path, &save_output)?;

    Ok("complete".to_string())
}
